OWML.ModHelper.Input/ModInputHandler.cs

```diff
@@ -0,0 +1,349 @@
+        private int _lastCombinationUpdate;
+        private readonly IModLogger _logger;
+        private readonly IModConsole _console;
+
+        internal bool ShouldIgnore(KeyCode code)
+        {
+            UpdateCurrentCombination();
+            return _currentCombination != null && Time.realtimeSinceStartup - _timeout[(int)code] < Cooldown;
+        }
+
+        public ModInputHandler(IModLogger logger, IModConsole console, IHarmonyHelper patcher)
+        {
+            _console = console;
+            _logger = logger;
+            patcher.AddPrefix<SingleAxisCommand>("ClearBinding", typeof(InputInterceptor), nameof(InputInterceptor.SingleAxisRemovePre));
+            patcher.AddPrefix<DoubleAxisCommand>("ClearBinding", typeof(InputInterceptor), nameof(InputInterceptor.DoubleAxisRemovePre));
+            patcher.AddPostfix<SingleAxisCommand>("Update", typeof(InputInterceptor), nameof(InputInterceptor.SingleAxisUpdatePost));
+            patcher.AddPostfix<DoubleAxisCommand>("Update", typeof(InputInterceptor), nameof(InputInterceptor.DoubleAxisUpdatePost));
+            Instance = this;
+        }
+
+        private long HashFromKeyboard()
+        {
+            long hash = 0;
+            var keysCount = 0;
+            var keys = new int[MaxComboLength];
+            var countdownTrigger = true;
+            for (var code = MinUsefulKey; code < MaxUsefulKey; code++)
+            {
+                if (!(Enum.IsDefined(typeof(KeyCode), (KeyCode)code) && UnityEngine.Input.GetKey((KeyCode)code)))
+                {
+                    continue;
+                }
+                keys[keysCount] = code;
+                keysCount++;
+                if (keysCount > MaxComboLength)
+                {
+                    return (int)RegistrationCode.CombinationTooLong;
+                }
+                hash = hash * MaxUsefulKey + code;
+                if (Time.realtimeSinceStartup - _timeout[code] > Cooldown)
+                {
+                    countdownTrigger = false;
+                }
+            }
+            return countdownTrigger ? -hash : hash;
+        }
+
+        private IModInputCombination CombinationFromKeyboard()
+        {
+            var countdownTrigger = false;
+            var hash = HashFromKeyboard();
+            if (hash < 0)
+            {
+                countdownTrigger = true;
+                hash = -hash;
+            }
+            if (!_comboRegistry.ContainsKey(hash))
+            {
+                return null;
+            }
+
+            var combination = _comboRegistry[hash];
+            if (!(combination == _currentCombination) && countdownTrigger)
+            {
+                return null;
+            }
+
+            if (hash < MaxUsefulKey)
+            {
+                return combination;
+            }
+            while (hash > 0)
+            {
+                _timeout[hash % MaxUsefulKey] = Time.realtimeSinceStartup;
+                hash /= MaxUsefulKey;
```

Comment:
Is this binary logic?

Reply:
hash is basically a polynomial or a number in 350 base

Reply:
implemented as a sorted array compressed into single 64 bit veriable, sort of a lossless hash